hat/app/org/hatdex/hat/api/controllers/ContractData.scala

```diff
@@ -0,0 +1,583 @@
+            handleJsValue(
+              hatUser.userId,
+              value,
+              dataEndpoint
+            )
+        }
+    }
+
+  // -- Update
+  def handleUpdateContractData(
+      hatUser: HatUser,
+      contractDataUpdate: ContractDataUpdateRequest,
+      namespace: String
+    )(implicit hatServer: HatServer): Future[Result] =
+    contractDataUpdate.body.length match {
+      // Missing Json Body, do nothing, return error
+      case 0 =>
+        logger.error(s"saveContractData included no json body - ns:${namespace}")
+        Future.successful(NotFound)
+      // There is a Json body, process either an JsArray or a JsValue
+      case _ =>
+        val dataSeq = contractDataUpdate.body.map(item => item.copy(endpoint = s"${namespace}/${item.endpoint}"))
+        dataService.updateRecords(
+          hatUser.userId,
+          dataSeq
+        ) map { saved =>
+          Created(Json.toJson(saved))
+        } recover {
+          case RichDataMissingException(message, _) =>
+            BadRequest(Json.toJson(ErrorsFromRichData.dataUpdateMissing(message)))
+        }
+    }
+
+  // --- API Handlers ---
+
+  def readContractData(
+      namespace: String,
+      endpoint: String,
+      orderBy: Option[String],
+      ordering: Option[String],
+      skip: Option[Int],
+      take: Option[Int]): Action[ContractDataReadRequest] =
+    UserAwareAction.async(parsers.json[ContractDataReadRequest]) { implicit request =>
+      val contractDataRead      = request.body
+      val maybeContractDataInfo = refineContractDataReadRequest(contractDataRead)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(_hatUser @ _), Right(RequestVerified(ns))) =>
+                makeData(ns, endpoint, orderBy, ordering, skip, take)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+    }
+
+  def createContractData(
+      namespace: String,
+      endpoint: String,
+      skipErrors: Option[Boolean]): Action[ContractDataCreateRequest] =
+    UserAwareAction.async(parsers.json[ContractDataCreateRequest]) { implicit request =>
+      val contractDataCreate    = request.body
+      val maybeContractDataInfo = refineContractDataCreateRequest(contractDataCreate)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(hatUser), Right(RequestVerified(ns))) =>
+                handleCreateContractData(hatUser, contractDataCreate, ns, endpoint, skipErrors)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
```

Comment:
As above is this complete, what will happen if the result is (None, Right(_))

Reply:
+1